activerecord/lib/active_record/reflection.rb

```diff
@@ -870,7 +870,7 @@ def association_class
       def association_primary_key(klass = nil)
         if primary_key = options[:primary_key]
           @association_primary_key ||= -primary_key.to_s
-        elsif !polymorphic? && ((klass || self.klass).has_query_constraints? || options[:query_constraints])
+        elsif (klass || self.klass).has_query_constraints? || options[:query_constraints]
```

Comment:
Hrm, I originally added the `!polymorphic?` check because `klass` can't always be computed on those associations, but this seems to be working fine based on the tests.